microbench/src/main/java/io/netty/microbench/websocket/WebSocketFrame08EncoderBenchmark.java

```diff
@@ -0,0 +1,99 @@
+ * under the License.
+ */
+package io.netty.microbench.websocket;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.websocketx.BinaryWebSocketFrame;
+import io.netty.handler.codec.http.websocketx.WebSocket08FrameEncoder;
+import io.netty.microbench.channel.EmbeddedChannelWriteReleaseHandlerContext;
+import io.netty.microbench.util.AbstractMicrobenchmark;
+import io.netty.util.internal.ThreadLocalRandom;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.profile.GCProfiler;
+import org.openjdk.jmh.runner.options.ChainedOptionsBuilder;
+
+@State(Scope.Benchmark)
+@Fork(value = 2)
+@Threads(1)
+@Warmup(iterations = 5)
+@Measurement(iterations = 10)
+public class WebSocketFrame08EncoderBenchmark extends AbstractMicrobenchmark {
+
+    private WebSocket08FrameEncoder websocketEncoder;
+
+    private ChannelHandlerContext context;
+
+    private ByteBuf content;
+
+    private BinaryWebSocketFrame webSocketFrame;
+    @Param({ "0", "4", "8", "32", "100", "1000", "3000" })
+    public int contentLength;
+
+    @Param({ "true", "false" })
+    public boolean pooledAllocator;
+
+    @Param({ "true" })
+    public boolean masking;
+
+    @Setup(Level.Trial)
+    public void setUp() {
+        byte[] bytes = new byte[contentLength];
+        ThreadLocalRandom.current().nextBytes(bytes);
+        ByteBufAllocator allocator = pooledAllocator? PooledByteBufAllocator.DEFAULT : UnpooledByteBufAllocator.DEFAULT;
+        content = allocator.buffer(contentLength).writeBytes(bytes);
+        ByteBuf testContent = Unpooled.unreleasableBuffer(content.asReadOnly());
+
+        webSocketFrame = new BinaryWebSocketFrame(testContent);
+        websocketEncoder = new WebSocket08FrameEncoder(masking);
+        context = new EmbeddedChannelWriteReleaseHandlerContext(allocator, websocketEncoder) {
+            @Override
+            protected void handleException(Throwable t) {
+                handleUnexpectedException(t);
+            }
+        };
+    }
+
+    @TearDown(Level.Trial)
+    public void teardown() {
+        content.release();
+        content = null;
+    }
+
+    @Benchmark
+    public void writeWebSocketFrame() throws Exception {
+        websocketEncoder.write(context, webSocketFrame, context.voidPromise());
```

Comment:
This benchmark may be wrong - https://github.com/openjdk/jmh/blob/master/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_08_DeadCode.java#L74

You need to consume some result or return it in this method, just to be sure you don't measure some side effects.

Reply:
I think it should work correct, because we not just convert frame or do simple computation which can be optimized away. We also use `handlerContext` where we accept converted frame and release it. However i return a `channelPromise` and rerun benchmarks.

Reply:
Checked benchmarks with returned promise it looks the same with small degradation.